README.md
## border_progress_indicator

##### Tiny package for indicating progress in a border.

![Image](https://raw.githubusercontent.com/berkekbgz/border_progress_indicator/main/resources/image.png)

### Usage

Wrap your widget in `BorderProgressIndicator`

```dart
BorderProgressIndicator(
	borderRadius: 10,
	value: 0.6,
	color: Colors.red,
	strokeWidth: 2,
	child: YOUR WIDGET HERE
),
```

#### Animated Version

You can also use self animated version with `AnimatedBorderProgressIndicator`.

```dart
AnimatedBorderProgressIndicator(
	borderRadius: 10,
	value: 0.6,
	color: Colors.red,
	curve: Curves.easeInCubic,
	strokeWidth: 2,
	child: YOUR WIDGET HERE
),
```



CHANGELOG.md
## 0.0.3

* Added example app

## 0.0.2

* Added example image to docs

## 0.0.1

* Initial release.


test/border_progress_indicator_test.dart
import 'package:flutter/material.dart';
import 'package:border_progress_indicator/border_progress_indicator.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    double value = .6;

    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              // Normal Version
              BorderProgressIndicator(
                borderRadius: 10,
                value: value,
                color: Colors.red,
                strokeWidth: 2,
                child: Container(
                  decoration: BoxDecoration(
                    color: Colors.grey.withOpacity(.5),
                    borderRadius: BorderRadius.circular(10),
                  ),
                  width: 200,
                  height: 50,
                  child: const Center(child: Text('Normal Version')),
                ),
              ),
              const SizedBox(height: 24),
              // Self Animated Version
              AnimatedBorderProgressIndicator(
                borderRadius: 10,
                value: value,
                color: Colors.red,
                curve: Curves.easeInCubic,
                strokeWidth: 2,
                child: Container(
                  decoration: BoxDecoration(
                    color: Colors.grey.withOpacity(.5),
                    borderRadius: BorderRadius.circular(10),
                  ),
                  width: 200,
                  height: 50,
                  child: const Center(child: Text('Animated Version')),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


lib/src/painter/border_painter.dart
import 'dart:ui';

import 'package:flutter/rendering.dart';

class BorderPainter extends CustomPainter {
  final double strokeWidth;
  final Color color;
  final double value;
  final double borderRadius;

  BorderPainter(
    this.value, {
    required this.color,
    this.strokeWidth = 2,
    this.borderRadius = 10,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final animationPercent = value;

    final path = Path()
      ..moveTo(size.width / 2, 0)
      ..lineTo(size.width - borderRadius, 0)
      ..arcToPoint(Offset(size.width, borderRadius),
          radius: Radius.circular(borderRadius))
      ..lineTo(size.width, size.height - borderRadius)
      ..arcToPoint(Offset(size.width - borderRadius, size.height),
          radius: Radius.circular(borderRadius))
      ..lineTo(borderRadius, size.height)
      ..arcToPoint(Offset(0, size.height - borderRadius),
          radius: Radius.circular(borderRadius))
      ..lineTo(0, borderRadius)
      ..arcToPoint(Offset(borderRadius, 0),
          radius: Radius.circular(borderRadius))
      ..lineTo(size.width / 2, 0);

    final totalLength = path
        .computeMetrics()
        .fold(0.0, (double prev, PathMetric metric) => prev + metric.length);

    final currentLength = totalLength * animationPercent;

    final followablePath = extractPathUntilLength(path, currentLength);

    final Paint paint = Paint();
    paint
      ..color = color
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth
      ..strokeCap = StrokeCap.round
      ..isAntiAlias = true
      ..filterQuality = FilterQuality.high
      ..strokeJoin = StrokeJoin.round;

    canvas.drawPath(followablePath, paint);
  }

  @override
  bool shouldRepaint(BorderPainter oldDelegate) => value != oldDelegate.value;

  Path extractPathUntilLength(
    Path originalPath,
    double length,
  ) {
    var currentLength = 0.0;

    final path = Path();

    var metricsIterator = originalPath.computeMetrics().iterator;

    while (metricsIterator.moveNext()) {
      var metric = metricsIterator.current;

      var nextLength = currentLength + metric.length;

      final isLastSegment = nextLength > length;
      if (isLastSegment) {
        final remainingLength = length - currentLength;
        final pathSegment = metric.extractPath(0.0, remainingLength);

        path.addPath(pathSegment, Offset.zero);
        break;
      } else {
        final pathSegment = metric.extractPath(0.0, metric.length);
        path.addPath(pathSegment, Offset.zero);
      }

      currentLength = nextLength;
    }

    return path;
  }
}


lib/src/border_progress_indicator.dart
import 'package:border_progress_indicator/src/painter/border_painter.dart';
import 'package:flutter/material.dart';

class BorderProgressIndicator extends StatefulWidget {
  final Widget? child;
  final double strokeWidth;
  final Color color;
  final double value;
  final double borderRadius;

  const BorderProgressIndicator({
    super.key,
    this.child,
    this.strokeWidth = 1,
    this.color = Colors.black,
    this.value = 0,
    this.borderRadius = 0,
  });

  @override
  State<BorderProgressIndicator> createState() =>
      _BorderProgressIndicatorState();
}

class _BorderProgressIndicatorState extends State<BorderProgressIndicator> {
  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: BorderPainter(
        widget.value,
        color: widget.color,
        strokeWidth: widget.strokeWidth,
        borderRadius: widget.borderRadius,
      ),
      child: widget.child,
    );
  }
}


lib/src/animated_border_progress_indicator.dart
import 'package:border_progress_indicator/src/painter/border_painter.dart';
import 'package:flutter/material.dart';

class AnimatedBorderProgressIndicator extends ImplicitlyAnimatedWidget {
  final Widget? child;
  final double strokeWidth;
  final Color color;
  final double value;
  final double borderRadius;

  const AnimatedBorderProgressIndicator({
    super.key,
    super.duration = const Duration(milliseconds: 300),
    super.curve = Curves.easeInOut,
    this.child,
    this.strokeWidth = 1,
    this.color = Colors.black,
    this.value = 0,
    this.borderRadius = 0,
  });

  @override
  AnimatedWidgetBaseState<AnimatedBorderProgressIndicator> createState() =>
      _AnimatedBorderProgressIndicatorState();
}

class _AnimatedBorderProgressIndicatorState
    extends AnimatedWidgetBaseState<AnimatedBorderProgressIndicator> {
  Tween<double>? _percent;
  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: BorderPainter(
        _percent!.evaluate(animation),
        color: widget.color,
        strokeWidth: widget.strokeWidth,
        borderRadius: widget.borderRadius,
      ),
      child: widget.child,
    );
  }

  @override
  void forEachTween(visitor) {
    _percent = visitor(
      _percent,
      widget.value,
      (dynamic value) => Tween<double>(begin: value as double),
    ) as Tween<double>;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:border_progress_indicator/border_progress_indicator.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    double value = .6;

    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text("${value * 100}%"),
              const SizedBox(height: 24),
              // Normal Version
              BorderProgressIndicator(
                borderRadius: 10,
                value: value,
                color: Colors.red,
                strokeWidth: 2,
                child: Container(
                  decoration: BoxDecoration(
                    color: Colors.grey.withOpacity(.5),
                    borderRadius: BorderRadius.circular(10),
                  ),
                  width: 200,
                  height: 50,
                  child: const Center(child: Text('Normal Version')),
                ),
              ),
              const SizedBox(height: 24),
              // Self Animated Version
              AnimatedBorderProgressIndicator(
                borderRadius: 10,
                value: value,
                color: Colors.red,
                curve: Curves.easeInCubic,
                strokeWidth: 2,
                child: SizedBox(
                  width: 200,
                  height: 50,
                  child: Container(
                    decoration: BoxDecoration(
                      color: Colors.grey.withOpacity(.5),
                      borderRadius: BorderRadius.circular(10),
                    ),
                    width: 200,
                    height: 50,
                    child: const Center(child: Text('Animated Version')),
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
